### PR TITLE
Add math/WPow

### PR DESCRIPTION
**Pull request by sblOWPCKCR**
+ WPow library to compute x^n
 + math.sh to test WPow. 2 types of tests: a couple of concrete ones (1^0, 2^2, 0^1, etc) and fuzz tests

### CODE REVIEW

File: test/math.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+set -euxo pipefail
+
+### test/fuzz test math utils (only WPow for the moment)
+
+# It's easy to do with dapptools (obviously), but we don't use it
+# Fortunately, it easy to make it work if you have dapptools installed
+
+# Step 1: prepare a temporary directory
+TMP=$(mktemp -d -t math)
+# (don't forget to clena after yourself)
```
Comment (alcueca) [inline test/math.sh:12]:
```suggestion
# (don't forget to clean after yourself)
```

File: test/WPow.t.sol [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            for (uint256 i = 1; i < n; ++i) {
+                unchecked {
+                    uint256 old_expected = expected;
+                    expected = (expected * x) / 1e18;
+                    // check for overflow if x > 1
+                    if (expectExpectedToIncrease && expected < old_expected) {
+                        return;
+                    }
+                    if (expected == old_expected) {
+                        break;
```
Comment (alcueca) [inline test/WPow.t.sol:39]: I guess that we get out of the loop because we have reached the precision limits.

File: test/WPow.t.sol [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        }
+        uint256 result = WPow.wpow(x, n);
+        uint256 distance = (result > expected)
+            ? result - expected
+            : expected - result;
+        emit log_uint(x);
+        emit log_uint(expected);
+        emit log_uint(result);
+        emit log_uint(distance);
+        uint acceptable_distance = result / 1e15;
```
Comment (alcueca) [inline test/WPow.t.sol:52]: Ok, so we only require 15 decimals of precision loss between the step-by-step calculation and `wpow`, right?

File: test/WPow.t.sol [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        assertLe(distance, acceptable_distance);
+    }
+
+    // pick a radom delta_x from [0, 2]
+    // pick a randrom n between 0 and 1000
+    // set x = 1 +/- delta_x
+    // check that wpow(x, y) is roughly the same as manual x * x * ... * x (n times)
+    // 'roughly the same': the diff is smaller than result * 1e-15
+    function testWPowBetween0and2(uint64 delta_x, uint16 n) public {
+        if (delta_x > 1e18) {
```
Comment (alcueca) [inline test/WPow.t.sol:66]: Shouldn't this be `delta_x > 2e18`?

File: test/WPow.t.sol [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            return;
+        }
+        if (n > 1000) {
+            return;
+        }
+        for (int256 sign = -1; sign <= 1; sign += 2) {
```
Comment (alcueca) [inline test/WPow.t.sol:72]: Why not replacing this block with:
```
testWPowImpl(x + delta_x, n);
testWPowImpl(x - delta_x, n);
```

File: test/WPow.t.sol [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        emit log_uint(distance);
+        uint acceptable_distance = result / 1e15;
+        // if x < 1e18, we're allowed to lose 1 wei of precision on each iteration
+        if (acceptable_distance < n) {
+            acceptable_distance = n;
+        }
+        assertLe(distance, acceptable_distance);
+    }
+
+    // pick a radom delta_x from [0, 2]
```
Comment (alcueca) [inline test/WPow.t.sol:60]:
```suggestion
    // pick a random delta_x from [0, 1]
```

File: test/WPow.t.sol [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    // pick a randrom n between 0 and 1000
+    // set x = 1 +/- delta_x
```
Comment (alcueca) [inline test/WPow.t.sol:62]:
```suggestion
    // set x = 1 +/- delta_x to test an x value from [0, 2]
```

**Review by alcueca — APPROVED** [review]
Approved, but please fix the comments first to make the fuzzing test a bit clearer.

File: test/WPow.t.sol (hunk 3 above)
Comment (sblOWPCKCR) [reply test/WPow.t.sol:52]: 3 decimals of loss (15 'correct' decimals). added the comment

File: test/WPow.t.sol (hunk 5 above)
Comment (sblOWPCKCR) [reply test/WPow.t.sol:72]: after 10 rewrites you have to have dumb code, right?

**Pull request merged**